### PR TITLE
👌 Improve MD comparison and testing

### PR DESCRIPTION
**Pull request by chrisjsewell**
1. Add plugin extensions to renderer before generating HTML 
2. Use HTML parsing to compare outputs; I think this is a bit more of an elegant solution
3. Use the markdown-it-py fixtures approach to testing input vs expected output rendering 😄 Its easier to add/modify tests, and serves as nice "specification" for the renderer that people can look at to quickly understand its behaviour (which is not as clear when looking at a JSON blob)
4. Add some extra tests for code formatter plugins

### CODE REVIEW

File: mdformat/_util.py [hunk 1]
```diff
@@ -1,21 +1,81 @@
...
+        plugin = mdformat.plugins.PARSER_EXTENSIONS[name]
+        plugin.update_mdit(mdit)
+        ignore_classes.extend(getattr(plugin, "ignore_classes", []))
+    html1 = HTML2JSON().parse(mdit.render(md1), ignore_classes)
+    html2 = HTML2JSON().parse(mdit.render(md2), ignore_classes)
+
     return html1 == html2
+
+
+class HTML2JSON(HTMLParser):
```
Comment (hukkinj1) [inline mdformat/_util.py:33]: This seems to output a syntax tree, not JSON. Maybe rename to `HTML2AST` or similar?

File: mdformat/renderer/_token_renderers.py [hunk 2]
```diff
@@ -101,9 +104,10 @@ def fence(tokens: List[Token], idx: int, options: dict, env: dict) -> str:
         fmt_func = options["codeformatters"][lang]
         try:
             code_block = fmt_func(code_block, info_str)
-        except Exception:
+        except Exception as err:
             # Swallow exceptions so that formatter errors (e.g. due to
             # invalid code) do not crash mdformat.
+            LOGGER.warning(f"Code formatting of '{lang}' failed: {err}")
             pass
```
Comment (hukkinj1) [inline mdformat/renderer/_token_renderers.py:111]: Remove this line?

File: mdformat/_util.py [hunk 3]
```diff
@@ -1,21 +1,81 @@
...
+            if self.current["tag"] == tag:
+                self.current = self.current.pop("parent")
+                break
+            self.current = self.current.pop("parent")
+
+    def handle_data(self, data: str) -> None:
+        if self.current:
+            self.current["data"] = data.rstrip()
+
+    def handle_comment(self, data: str) -> None:
```
Comment (hukkinj1) [inline mdformat/_util.py:78]: Do we need to define this? Maybe just skip comments?

File: mdformat/_util.py [hunk 4]
```diff
@@ -1,21 +1,81 @@
...
+    def handle_endtag(self, tag: str) -> None:
+        # walk up the tree to the tag's parent
+        while self.current is not None:
+            if self.current["tag"] == tag:
+                self.current = self.current.pop("parent")
+                break
+            self.current = self.current.pop("parent")
+
+    def handle_data(self, data: str) -> None:
+        if self.current:
```
Comment (hukkinj1) [inline mdformat/_util.py:75]: Can `self.current` ever be falsy, given valid HTML? If not, maybe remove the check and fail fast and hard?

File: mdformat/_util.py [hunk 5]
```diff
@@ -1,21 +1,81 @@
...
-    html2 = re.sub(r"\s+", "", html2)
-    for codeclass in ignore_codeclasses:
-        html1 = re.sub(rf'<codeclass="language-{codeclass}">.*</pre>', "", html1)
-        html2 = re.sub(rf'<codeclass="language-{codeclass}">.*</pre>', "", html2)
+    ignore_classes = [f"language-{lang}" for lang in enabled_codeformatters]
+    mdit = MarkdownIt()
+    for name in enabled_extensions:
+        plugin = mdformat.plugins.PARSER_EXTENSIONS[name]
+        plugin.update_mdit(mdit)
+        ignore_classes.extend(getattr(plugin, "ignore_classes", []))
```
Comment (hukkinj1) [inline mdformat/_util.py:26]: This is new plugin API and should be documented if we leave it in. What would this be used for though? There aren't many classes in HTML generated from Markdown.

I've been thinking to make a TOC generator plugin where I would need to ignore everything in the HTML that's in between certain types of HTML comments. A regex based ignore system could work there, or a flag that disables equality check altogether, but not sure if a class based system like this would.

**Review by hukkinj1 — CHANGES_REQUESTED** [review]

File: mdformat/_util.py [hunk 6]
```diff
@@ -1,21 +1,81 @@
...
+            self.tree.append(tag_item)
+        else:
+            children = self.current.setdefault("children", [])
+            children.append(tag_item)
+        self.current = tag_item
+
+    def handle_endtag(self, tag: str) -> None:
+        # walk up the tree to the tag's parent
+        while self.current is not None:
+            if self.current["tag"] == tag:
```
Comment (hukkinj1) [inline mdformat/_util.py:69]: Is it possible that this exits too early here, if say a `<ul>` tag is nested inside another `<ul>` tag?

File: mdformat/_util.py (hunk 1 above)
Comment (chrisjsewell) [reply mdformat/_util.py:33]: Well yes it does produce valid JSON, which also happens to be an AST of sorts. So really either is apt.
Note that the other approach here is just to use https://www.crummy.com/software/BeautifulSoup/bs4/doc/ (that's the code I originally wrote), i.e. having an extra (albeit very well known) dependency versus using a bespoke parser that requires a little more validation testing. I'll leave you to decide which?

File: mdformat/renderer/_token_renderers.py (hunk 2 above)
Comment (chrisjsewell) [reply mdformat/renderer/_token_renderers.py:111]: If we are addressing later it in #37 then yeh sure

File: mdformat/_util.py (hunk 3 above)
Comment (chrisjsewell) [reply mdformat/_util.py:78]: yeh I guess it just depends on how rigorous you want to be. Happy to drop

File: mdformat/_util.py (hunk 4 above)
Comment (chrisjsewell) [reply mdformat/_util.py:75]: well self.current can technically be `None` (although it will have been replaced in `handle_starttag` by this point) and so mypy complained about it

File: mdformat/renderer/_token_renderers.py (hunk 2 above)
Comment (hukkinj1) [reply mdformat/renderer/_token_renderers.py:111]: I refer to line 111, not 110 :smile:

File: mdformat/_util.py (hunk 5 above)
Comment (chrisjsewell) [reply mdformat/_util.py:26]: why in between HTML comments? what would the ToC HTML structure look like?
and anyway, if we drop `handle_comment` (above), then comments would be ignored anyway?

well you could have something like the amsmath plugin that generates:

```html
<section class="amsmath">
<eqn>
\begin{equation*}
a = 1
\end{equation*}
</eqn>
</section>
```

and so if you wanted to format the TeX in some way, you would specify `ignore_classes = ["amsmath"]`. You could even take it one step further and specify CSS selector like syntax, e.g. `section.amsmath` (this is available in beautifulsoup)

File: mdformat/renderer/_token_renderers.py (hunk 2 above)
Comment (chrisjsewell) [reply mdformat/renderer/_token_renderers.py:111]: oh yeh oops

File: mdformat/_util.py (hunk 3 above)
Comment (hukkinj1) [reply mdformat/_util.py:78]: Comments can't have any effect on rendered HTML so IMO we don't need to care about them here.

File: mdformat/renderer/_token_renderers.py (hunk 2 above)
Comment (chrisjsewell) [reply mdformat/renderer/_token_renderers.py:111]: we can probably improve on this anyway, but will leave it in there for a start

File: mdformat/_util.py (hunk 1 above)
Comment (hukkinj1) [reply mdformat/_util.py:33]: I believe JSON is a text format, so we'd need to do `json.dumps()` to the output before we can call it JSON.

I think this custom parser should be just fine if we can keep it reasonably compact.

Comment (chrisjsewell) [reply mdformat/_util.py:33]: yep fair

File: mdformat/_util.py (hunk 4 above)
Comment (hukkinj1) [reply mdformat/_util.py:75]: Aah if its just mypy complaining, I'd help mypy with a `assert self.current is not None` instead of the `if` clause

File: mdformat/_util.py (hunk 6 above)
Comment (chrisjsewell) [reply mdformat/_util.py:69]: No pretty sure not, because it first exits the inner </ul> then the outer `</ul>`.
I can add an extra test though.

File: mdformat/_util.py (hunk 5 above)
Comment (chrisjsewell) [reply mdformat/_util.py:26]: I can add it to the `ParserExtensionInterface` though

Comment (hukkinj1) [reply mdformat/_util.py:26]: > what would the ToC HTML structure look like?

Something like https://github.com/thlorenz/doctoc#specifying-location-of-toc

So you leave these comments in the Markdown and the tool will generate a TOC in between them.

Comment (hukkinj1) [reply mdformat/_util.py:26]: Hmm, I'm thinking maybe just a boolean switch to turn off the Markdown equality check altogether might be the cleanest solution that would work for every possible use case, and with very little effort...?

Comment (chrisjsewell) [reply mdformat/_util.py:26]: yeh that feels like a bit of an edge case that we can deal with later

**Comment by chrisjsewell** [comment]
applied changes

File: mdformat/_util.py (hunk 6 above)
Comment (chrisjsewell) [reply mdformat/_util.py:69]: added test

File: mdformat/_util.py [hunk 7]
```diff
@@ -1,21 +1,78 @@
...
+        plugin.update_mdit(mdit)
+        ignore_classes.extend(getattr(plugin, "ignore_classes", []))
+    html1 = HTML2AST().parse(mdit.render(md1), ignore_classes)
+    html2 = HTML2AST().parse(mdit.render(md2), ignore_classes)
+
     return html1 == html2
+
+
+class HTML2AST(HTMLParser):
+    """Parser HTML to JSON."""
```
Comment (hukkinj1) [inline mdformat/_util.py:34]: JSON --> AST rename

File: mdformat/_util.py [hunk 8]
```diff
@@ -1,21 +1,78 @@
...
+            if self.current["tag"] == tag:
+                self.current = self.current.pop("parent")
+                break
+            self.current = self.current.pop("parent")
+
+    def handle_data(self, data: str) -> None:
+        # ignore empty data
+        if data.strip():
+            assert self.current is not None, f"'{data}'"
+            self.current["data"] = data.rstrip()
```
Comment (hukkinj1) [inline mdformat/_util.py:78]: Is there a reason why `strip` above and `rstrip` here? Maybe `strip` both?

File: mdformat/_util.py (hunk 5 above)
Comment (chrisjsewell) [reply mdformat/_util.py:26]: BTW it feels like there could be some possible integration of the ToC plugin and https://github.com/executablebooks/markdown-it-py/blob/master/markdown_it/extensions/anchors/index.py

Comment (chrisjsewell) [reply mdformat/_util.py:26]: i.e. potentially moving the anchor plugin code out of markdown-it-py and into the same package as the toc generator plugin

Comment (hukkinj1) [reply mdformat/_util.py:26]: Ah so what I'm suggesting already now, before we add the class-based-skip-equality-check API, should we just instead give plugins a boolean flag they can set that completely disables the equality check. The ToC plugin for instance seems to need something more powerful than the class-based system, and I wouldn't want to end up supporting many ways to bypass the equality check.

This way:
- the API is dead simple
- we never have to extend it

The whole equality check is just a safeguard really, intended for the default "pure" formatter mode, that is not supposed to alter AST.

**Review by hukkinj1 — CHANGES_REQUESTED** [review]

File: mdformat/_util.py (hunk 7 above)
Comment (chrisjsewell) [reply mdformat/_util.py:34]: done

File: mdformat/_util.py (hunk 8 above)
Comment (chrisjsewell) [reply mdformat/_util.py]: I've gone back to ignoring data outside tags (i.e. ignore `self.current = None`) so that it ignores any newline markers at the start/end of the file.
I've also changed to splitting into lines and removing all empty lines and trailing whitespace.
I think this is good enough to check for a high degree of equality.

File: mdformat/_util.py (hunk 5 above)
Comment (chrisjsewell) [reply mdformat/_util.py:26]: I prefer to keep the equality check where possible. I think for most plugins the `ignore_classes` option should be sufficient.
This ToC plugin feels like a bit of special case, in that it is actually adding something new to the document, rather than just formatting what is already there

Comment (hukkinj1) [reply mdformat/_util.py:26]: Ok so I thought about this a bit. Current thoughts:

- I will want to add a `CHANGES_AST` (or similar name) switch to the plugin API in any case.
- The equality check is really just a safeguard for the CLI (currently not applied to Python API), not a core part of mdformat. Assuming no bugs, the feature does nothing. I want to avoid overengineering a set of plugin hooks to alter that.
- The formatter in theory should never alter AST/rendered markdown. I feel like any plugin that does so, is equally a special case.

So I'm thinking to reject the class-based plugin hook. I know this could be a bit frustrating, sorry for that. And if so, I encourage to plan new features in an issue before implementation.

Comment (chrisjsewell) [reply mdformat/_util.py:26]: could the equality check be toggled to only check the equality of the HTML AST structure, rather than its full content, i.e. `def handle_data` would not store anything in this mode.
This would still check that the formatting does not change the overall tag structure of the HTML (which a plugin should not generally do), but would not worry about its content (which mdformat core and plugins literally all change).

e.g. it would just check that both docs have the structure

```html
<div a="1">
  <other b="2">
  </other>
</div>
```

rather than check that they are exactly:

```html
<div a="1">
  a b
  c d
  <other b="2">
  x y z
  </other>
</div>
```

Comment (chrisjsewell) [reply mdformat/_util.py:26]: > The formatter in theory should never alter AST/rendered markdown

and yeh just to emphasise; in some sense the formatter *always* alters the HTML AST, if you are taking into account the content of each tag, e.g. stripping whitespace in this context is altering the AST.

what the formatter should not do (in theory) is alter the HTML AST structure

Comment (hukkinj1) [reply mdformat/_util.py:26]: > could the equality check be toggled to only check the equality of the HTML AST structure, rather than its full content

This would also require an addition to the API, which I'd rather not do. I've thought about possibly even removing the equality check once mdformat proves to be stable and bug free (since it makes the tool render 3 docs instead of just 1). I'd prefer the API to not be littered with stuff related.

> This would still check that the formatting does not change the overall tag structure of the HTML (which a plugin should not generally do), but would not worry about its content (which mdformat core and plugins literally all change).

> stripping whitespace in this context is altering the AST.

I mean CommonMark AST, which AFAIK should completely ignore leading or trailing whitespace. The HTML rendering is just a tool used to check that.
Besides leading/trailing whitespace the changes mdformat does are mostly just adding escape characters and converting characters to their numeric representation (which AFAIK should not change HTML in any way).

> what the formatter should not do (in theory) is alter the HTML AST structure

It's more strict than this. All ASTs aside, what I really want is the rendered HTML to look precisely the same before and after running mdformat (core). That means that the changes mentioned above are fine (when applied to paragraphs), but most other changes to content (e.g. any changes to code blocks) are not. Plugins are free to do what they want of course.

Comment (chrisjsewell) [reply mdformat/_util.py:26]: > I've thought about possibly even removing the equality check

I don't know, I quite like it. You probably already know, but black has a similar equivalence check
https://github.com/psf/black/blob/172c0a78facd8f938629f216c95507bbfeddfe5e/src/black/__init__.py#L944

> I mean CommonMark AST, which AFAIK should completely ignore leading or trailing whitespace.

I'm not sure if this is correct. I mean there's not really a CommonMark AST anyway, just a specification of what HTML it should produce.

It would just be nice if you didn't have to completely throw away this check, due to the use of one plugin.

Comment (hukkinj1) [reply mdformat/_util.py:26]: > I'm not sure if this is correct. I mean there's not really a CommonMark AST anyway, just a specification of what HTML it should produce.

I'm pretty sure every language can be parsed into a syntax tree, even if there's not an established format how it should look like. According to CommonMark spec `The paragraph’s raw content is formed by concatenating the lines and removing initial and final whitespace.`, so a parser will definitely remove leading/trailing whitespace and it won't be in this syntax tree.

> It would just be nice if you didn't have to completely throw away this check, due to the use of one plugin.

With the `CHANGES_AST` flag, you don't have throw it away, unless the plugin makes changes to how the rendered Markdown looks like. That seems like simple and reasonable spec to me.

I think at this point we should just agree to disagree? :smile: :peace_symbol:

**Comment by hukkinj1** [comment]
I will close this PR as I don't want the class based equality check ignore. Without it I feel like the HTML parser might also not be necessary. The tests here I definitely want though. Many of the test changes, and the warning log added here seem to already be in `master`.

I've taken the commits, applied changes from `master` and cut out the changes that seem unnecessary. The result is here https://github.com/executablebooks/mdformat/pull/66

The tests found one bug in `master`, thank you very much @chrisjsewell !
